File: apps/api-rs/src/rag/reranker.rs
use anyhow::{anyhow, bail, Result};
use serde::{Deserialize, Serialize};

use crate::models::rag::{RerankInput, RerankedChunk};

#[async_trait::async_trait]
pub trait Reranker: Send + Sync {
    async fn rerank(&self, input: RerankInput) -> Result<Vec<RerankedChunk>>;
    fn component_name(&self) -> String;
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RerankProvider {
    DashScope,
    OpenAiCompatible,
}

impl RerankProvider {
    pub fn parse(value: &str) -> Result<Self> {
        match value.trim().to_ascii_lowercase().as_str() {
            "dashscope" => Ok(Self::DashScope),
            "openai" | "openai_compatible" | "cohere_compatible" => Ok(Self::OpenAiCompatible),
            other => bail!("unsupported rerank provider: {other}"),
        }
    }

    fn as_str(self) -> &'static str {
        match self {
            Self::DashScope => "dashscope",
            Self::OpenAiCompatible => "openai_compatible",
        }
    }
}

pub struct HttpReranker {
    http: reqwest::Client,
    api_url: String,
    api_key: Option<String>,
    model: String,
    provider: RerankProvider,
}

#[derive(Debug, Serialize)]
struct CompatibleRequest<'a> {
    model: &'a str,
    query: &'a str,
    documents: &'a [String],
    top_n: usize,
}

#[derive(Debug, Serialize)]
struct DashScopeRequest<'a> {
    model: &'a str,
    input: DashScopeInput<'a>,
    parameters: DashScopeParameters,
}

#[derive(Debug, Serialize)]
struct DashScopeInput<'a> {
    query: &'a str,
    documents: &'a [String],
}

#[derive(Debug, Serialize)]
struct DashScopeParameters {
    return_documents: bool,
    top_n: usize,
}

#[derive(Debug, Deserialize)]
struct RerankResult {
    index: usize,
    #[serde(default, alias = "relevance_score")]
    score: f64,
}

impl HttpReranker {
    pub fn new(
        api_url: String,
        api_key: Option<String>,
        model: String,
        provider: RerankProvider,
    ) -> Result<Self> {
        if api_url.trim().is_empty() {
            bail!("rerank api url is empty");
        }
        if model.trim().is_empty() {
            bail!("rerank model is empty");
        }
        let http = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(60))
            .build()?;
        Ok(Self {
            http,
            api_url,
            api_key,
            model,
            provider,
        })
    }

    async fn request(
        &self,
        query: &str,
        documents: &[String],
        top_n: usize,
    ) -> Result<Vec<RerankResult>> {
        let mut request = self
            .http
            .post(&self.api_url)
            .header("Content-Type", "application/json");
        if let Some(api_key) = &self.api_key {
            request = request.bearer_auth(api_key);
        }
        request = match self.provider {
            RerankProvider::DashScope => request.json(&DashScopeRequest {
                model: &self.model,
                input: DashScopeInput { query, documents },
                parameters: DashScopeParameters {
                    return_documents: false,
                    top_n,
                },
            }),
            RerankProvider::OpenAiCompatible => request.json(&CompatibleRequest {
                model: &self.model,
                query,
                documents,
                top_n,
            }),
        };
        let response = request.send().await?;
        let status = response.status();
        let body = response.text().await?;
        if !status.is_success() {
            let diagnostic = body.chars().take(1_000).collect::<String>();
            bail!("reranker request failed with HTTP {status}: {diagnostic}");
        }
        let payload: serde_json::Value = serde_json::from_str(&body)
            .map_err(|error| anyhow!("reranker returned invalid JSON: {error}"))?;
        parse_results(&payload)
    }

    pub async fn probe(&self) -> Result<()> {
        let results = self
            .request(
                "enterprise document retrieval readiness",
                &[
                    "enterprise document retrieval readiness".to_string(),
                    "unrelated weather observation".to_string(),
                ],
                1,
            )
            .await?;
        let first = results
            .first()
            .ok_or_else(|| anyhow!("reranker readiness probe returned no result"))?;
        if first.index >= 2 || !first.score.is_finite() {
            bail!("reranker readiness probe returned an invalid result");
        }
        Ok(())
    }
}

#[async_trait::async_trait]
impl Reranker for HttpReranker {
    async fn rerank(&self, input: RerankInput) -> Result<Vec<RerankedChunk>> {
        if input.chunks.is_empty() {
            return Ok(vec![]);
        }
        let documents = input
            .chunks
            .iter()
            .map(rerank_document_text)
            .collect::<Vec<_>>();
        let mut results = self
            .request(&input.query, &documents, input.top_k.max(1))
            .await?;
        results.sort_by(|a, b| {
            b.score
                .partial_cmp(&a.score)
                .unwrap_or(std::cmp::Ordering::Equal)
        });

        let mut reranked = Vec::new();
        for result in results.into_iter().take(input.top_k.max(1)) {
            let chunk = input.chunks.get(result.index).cloned().ok_or_else(|| {
                anyhow!("reranker returned invalid document index {}", result.index)
            })?;
            reranked.push(RerankedChunk {
                chunk,
                score: result.score,
                rank: reranked.len() as i32 + 1,
            });
        }
        Ok(reranked)
    }

    fn component_name(&self) -> String {
        format!("{}:{}", self.provider.as_str(), self.model)
    }
}

fn parse_results(payload: &serde_json::Value) -> Result<Vec<RerankResult>> {
    let results = payload
        .get("results")
        .or_else(|| {
            payload
                .get("output")
                .and_then(|output| output.get("results"))
        })
        .ok_or_else(|| anyhow!("reranker response is missing results"))?;
    let parsed = serde_json::from_value::<Vec<RerankResult>>(results.clone())?;
    if parsed.is_empty() {
        bail!("reranker returned no results");
    }
    Ok(parsed)
}

fn rerank_document_text(chunk: &crate::models::rag::RetrievedChunk) -> String {
    serde_json::json!({
        "document_title": chunk.doc_title,
        "heading_path": chunk.heading_path,
        "content": chunk.content,
    })
    .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_dashscope_response() {
        let value = serde_json::json!({
            "output": {"results": [
                {"index": 1, "relevance_score": 0.91},
                {"index": 0, "relevance_score": 0.22}
            ]}
        });
        let results = parse_results(&value).expect("dashscope response should parse");
        assert_eq!(results.len(), 2);
        assert_eq!(results[0].index, 1);
        assert!((results[0].score - 0.91).abs() < f64::EPSILON);
    }

    #[test]
    fn rejects_unknown_provider() {
        assert!(RerankProvider::parse("rule_based").is_err());
    }
}

File: apps/api-bun/src/rag/reranker.ts
// 移植自 apps/api-rs/src/rag/reranker.rs —— HttpReranker（dashscope / openai 兼容两种请求体）
import type { RerankInput, RerankedChunk, RetrievedChunk } from '../models/rag.ts';
import type { Reranker, RerankProviderKind } from './types.ts';
import { asRecord } from './retriever/es_source.ts';

export type RerankProvider = 'dashscope' | 'openai_compatible';

export function parseRerankProvider(value: string): RerankProvider {
  switch (value.trim().toLowerCase()) {
    case 'dashscope':
      return 'dashscope';
    case 'openai':
    case 'openai_compatible':
    case 'cohere_compatible':
      return 'openai_compatible';
    default:
      throw new Error(`unsupported rerank provider: ${value.trim().toLowerCase()}`);
  }
}

/** 将 types.ts 的具体厂商 kind 映射到 Rust 的两种请求形态。 */
export function rerankProviderFromKind(kind: RerankProviderKind): RerankProvider {
  switch (kind) {
    case 'dashscope':
      return 'dashscope';
    case 'jina':
    case 'cohere':
    case 'siliconflow':
      return 'openai_compatible';
  }
}

function providerAsStr(provider: RerankProvider): string {
  return provider;
}

interface RerankResult {
  index: number;
  score: number;
}

const RERANK_TIMEOUT_SECONDS = 60;

export class HttpReranker implements Reranker {
  private readonly apiUrl: string;
  private readonly apiKey: string | null;
  private readonly model: string;
  private readonly provider: RerankProvider;

  constructor(apiUrl: string, apiKey: string | null, model: string, provider: RerankProvider) {
    if (apiUrl.trim().length === 0) throw new Error('rerank api url is empty');
    if (model.trim().length === 0) throw new Error('rerank model is empty');
    this.apiUrl = apiUrl;
    this.apiKey = apiKey;
    this.model = model;
    this.provider = provider;
  }

  componentName(): string {
    return `${providerAsStr(this.provider)}:${this.model}`;
  }

  private async request(query: string, documents: string[], topN: number): Promise<RerankResult[]> {
    const headers: Record<string, string> = { 'Content-Type': 'application/json' };
    if (this.apiKey !== null) headers.Authorization = `Bearer ${this.apiKey}`;
    const body =
      this.provider === 'dashscope'
        ? JSON.stringify({
            model: this.model,
            input: { query, documents },
            parameters: { return_documents: false, top_n: topN },
          })
        : JSON.stringify({ model: this.model, query, documents, top_n: topN });
    const response = await fetch(this.apiUrl, {
      method: 'POST',
      headers,
      body,
      signal: AbortSignal.timeout(RERANK_TIMEOUT_SECONDS * 1000),
    });
    const status = response.status;
    const text = await response.text();
    if (!response.ok) {
      const diagnostic = Array.from(text).slice(0, 1_000).join('');
      throw new Error(`reranker request failed with HTTP ${status} ${response.statusText}: ${diagnostic}`);
    }
    let payload: unknown;
    try {
      payload = JSON.parse(text);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new Error(`reranker returned invalid JSON: ${message}`);
    }
    return parseResults(payload);
  }

  async probe(): Promise<void> {
    const results = await this.request(
      'enterprise document retrieval readiness',
      [
        'enterprise document retrieval readiness',
        'unrelated weather observation',
      ],
      1,
    );
    const first = results[0];
    if (first === undefined) {
      throw new Error('reranker readiness probe returned no result');
    }
    if (first.index >= 2 || !Number.isFinite(first.score)) {
      throw new Error('reranker readiness probe returned an invalid result');
    }
  }

  async rerank(input: RerankInput): Promise<RerankedChunk[]> {
    if (input.chunks.length === 0) return [];
    const documents = input.chunks.map(rerankDocumentText);
    const results = await this.request(input.query, documents, Math.max(1, input.top_k));
    results.sort((a, b) => compareDesc(a.score, b.score));

    const reranked: RerankedChunk[] = [];
    for (const result of results.slice(0, Math.max(1, input.top_k))) {
      const chunk = input.chunks[result.index];
      if (chunk === undefined) {
        throw new Error(`reranker returned invalid document index ${result.index}`);
      }
      reranked.push({ chunk, score: result.score, rank: reranked.length + 1 });
    }
    return reranked;
  }
}

/** 与 Rust partial_cmp(...).unwrap_or(Equal) 语义一致：NaN 视为相等。 */
function compareDesc(left: number, right: number): number {
  if (Number.isNaN(left) || Number.isNaN(right)) return 0;
  if (right > left) return 1;
  if (right < left) return -1;
  return 0;
}

export function parseResults(payload: unknown): RerankResult[] {
  const record = asRecord(payload);
  if (record === null) throw new Error('reranker response is missing results');
  let results = record.results;
  if (results === undefined || results === null) {
    const output = asRecord(record.output);
    results = output === null ? undefined : output.results;
  }
  if (!Array.isArray(results)) throw new Error('reranker response is missing results');
  const parsed: RerankResult[] = [];
  for (const item of results) {
    const itemRecord = asRecord(item);
    if (itemRecord === null) throw new Error('reranker returned an invalid result item');
    const index = itemRecord.index;
    if (typeof index !== 'number' || !Number.isInteger(index)) {
      throw new Error('reranker returned an invalid result item');
    }
    const scoreValue = itemRecord.score ?? itemRecord.relevance_score ?? 0;
    if (typeof scoreValue !== 'number') throw new Error('reranker returned an invalid result item');
    parsed.push({ index, score: scoreValue });
  }
  if (parsed.length === 0) throw new Error('reranker returned no results');
  return parsed;
}

function rerankDocumentText(chunk: RetrievedChunk): string {
  return JSON.stringify({
    document_title: chunk.doc_title,
    heading_path: chunk.heading_path,
    content: chunk.content,
  });
}
